src/cdk/app.ts

```diff
@@ -14,43 +13,140 @@ export class LambdaService extends cdk.Stack {
         });
 
-        const restApiName = `contributions-service-${stage.value}`;
+        const vpc = ec2.Vpc.fromVpcAttributes(this, 'vpc', {
+            vpcId: vpcId.valueAsString,
+            availabilityZones: availabilityZones.valueAsList,
+            publicSubnetIds: publicSubnets.valueAsList,
+        });
+
+        const role = new iam.Role(this, 'role', {
+            assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
+            inlinePolicies: {
+                instancePolicy: new iam.PolicyDocument({
+                    statements: [
+                        new iam.PolicyStatement({
+                            effect: iam.Effect.ALLOW,
+                            resources: ['arn:aws:s3:::aws-frontend-contributions-service/*'],
+                            actions: ['s3:GetObject', 's3:HeadObject', 's3:List*'],
+                        }),
+                        new iam.PolicyStatement({
+                            effect: iam.Effect.ALLOW,
+                            resources: ['*'],
+                            actions: ['ssmmessages:*', 'ssm:*', 'ec2messages:*', 'logs:*'],
+                        }),
+                    ],
+                }),
+            },
+        });
+
+        const userData = ec2.UserData.forLinux();
+        const App = app.value.toString();
+        const Stack = stack.value.toString();
+        const Stage = stage.value.toString();
+
+        userData.addCommands(
+            `groupadd frontend`,
+            `useradd -r -m -s /usr/bin/nologin -g frontend ${App}`,
+
+            `export App=${App}`,
+            `export Stack=${Stack}`,
+            `export Stage=${Stage}`,
+            `export NODE_ENV=production`,
+            `export BASE_URL=\`aws ssm get-parameter --name "/contributions-service/${Stage.toLowerCase()}/base_url" --region eu-west-1 | jq ".Parameter.Value"\``,
+            `export LOG_TARGETING=\`aws ssm get-parameter --name "/contributions-service/${Stage.toLowerCase()}/log_targeting" --region eu-west-1 | jq ".Parameter.Value"\``,
+            `export LOG_COMPARE_VARIANTS=\`aws ssm get-parameter --name "/contributions-service/${Stage.toLowerCase()}/log_compare_variants" --region eu-west-1 | jq ".Parameter.Value"\``,
+            `export LOG_FAILED_TEST_FILTER=\`aws ssm get-parameter --name "/contributions-service/${Stage.toLowerCase()}/log_failed_test_filter" --region eu-west-1 | jq ".Parameter.Value"\``,
+
+            `aws s3 cp s3://aws-frontend-contributions-service/frontend/${Stage}/contributions-service-ec2/contributions-service-ec2.zip /tmp/${App}.zip`,
+            `mkdir -p /opt/${App}`,
+            `unzip /tmp/${App}.zip -d /opt/${App}`,
+            `chown -R ${App}:frontend /opt/${App}`,
+
+            `export PM2_HOME="/usr/share/${App}"`,
+
+            `/usr/local/node/pm2 start --name ${App} --uid ${App} --gid frontend /opt/${App}/server.js`,
+        );
+
+        const asg = new autoscaling.AutoScalingGroup(this, 'ASG', {
```

Comment:
TODO: confirm the values (instance sizes, number of instances) to use for the ASG.